File: php/challenge_0.php
<?php
error_reporting(E_ALL);
ini_set('display_errors', 1);

// Server configuration
$host = "0.0.0.0";
$port = 40000;

// Create a TCP socket
$socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
if ($socket === false) {
    die("Socket creation failed: " . socket_strerror(socket_last_error()));
}

// Set socket options
socket_set_option($socket, SOL_SOCKET, SO_REUSEADDR, 1);

// Bind to host and port
if (!socket_bind($socket, $host, $port)) {
    die("Socket bind failed: " . socket_strerror(socket_last_error($socket)));
}

// Listen for connections
if (!socket_listen($socket, 5)) {
    die("Socket listen failed: " . socket_strerror(socket_last_error($socket)));
}

echo "Server listening on $host:$port\n";

// Accept connections in a loop
while (true) {
    $client = socket_accept($socket);
    if ($client === false) {
        echo "Socket accept failed: " . socket_strerror(socket_last_error($socket)) . "\n";
        continue;
    }

    // Handle each client in a separate process
    $pid = pcntl_fork();
    if ($pid == -1) {
        die("Could not fork process\n");
    } elseif ($pid) {
        // Parent process: close client socket and continue accepting
        socket_close($client);
    } else {
        // Child process: handle client communication
        while (true) {
            $data = socket_read($client, 1024, PHP_BINARY_READ);
            if ($data === false || $data === "") {
                break; // Client disconnected or error
            }
            socket_write($client, $data, strlen($data));
        }
        socket_close($client);
        exit(0); // Exit child process
    }
}

// Close server socket (unreachable in this loop, but good practice)
socket_close($socket);

File: Pascal/challenge_0.pas
program SmokeTest;

{$mode objfpc}{$H+}

uses
  Sockets, SysUtils, BaseUnix;  // BaseUnix for ESysEINTR

const
  BUFFER_SIZE = 1024;

var
  ServerSocket, ClientSocket: Integer;
  Buffer: array[0..BUFFER_SIZE-1] of Byte;
  BytesRead: LongInt;

function CreateServer(Port: Word): Integer;
var
  Sock: Integer;
  Opt: LongInt;
  Addr: TInetSockAddr;
begin
  // Create TCP socket
  Sock := fpSocket(AF_INET, SOCK_STREAM, 0);
  if Sock = -1 then
  begin
    WriteLn('Error creating socket: ', GetLastOSError);
    Halt(1);
  end;

  // Allow socket to reuse address
  Opt := 1;
  if fpSetSockOpt(Sock, SOL_SOCKET, SO_REUSEADDR, @Opt, SizeOf(Opt)) = -1 then
  begin
    WriteLn('Error setting socket options: ', GetLastOSError);
    CloseSocket(Sock);
    Halt(1);
  end;

  // Bind socket to address and port
  FillChar(Addr, SizeOf(Addr), 0);
  Addr.sin_family := AF_INET;
  Addr.sin_port   := htons(Port);
  Addr.sin_addr.s_addr := htonl(INADDR_ANY);

  if fpBind(Sock, @Addr, SizeOf(Addr)) = -1 then
  begin
    WriteLn('Error binding socket: ', GetLastOSError);
    CloseSocket(Sock);
    Halt(1);
  end;

  // Listen for incoming connections
  if fpListen(Sock, 5) = -1 then
  begin
    WriteLn('Error listening on socket: ', GetLastOSError);
    CloseSocket(Sock);
    Halt(1);
  end;

  Result := Sock;
end;

procedure HandleClient(ClientSock: Integer);
var
  BytesSent: LongInt;
begin
  while True do
  begin
    // Receive data from client
    BytesRead := fpRecv(ClientSock, @Buffer, BUFFER_SIZE, 0);
    if BytesRead <= 0 then
    begin
      if BytesRead = -1 then
        WriteLn('Error receiving data: ', GetLastOSError);
      Break; // Client disconnected or error
    end;

    // Echo back the received data
    BytesSent := fpSend(ClientSock, @Buffer, BytesRead, 0);
    if BytesSent = -1 then
    begin
      WriteLn('Error sending data: ', GetLastOSError);
      Break;
    end;
  end;

  // Close client socket
  CloseSocket(ClientSock);
end;

begin
  // Create server on port 40000 (change if needed)
  ServerSocket := CreateServer(40000);
  WriteLn('Server listening on port 40000...');

  // Main server loop
  while True do
  begin
    // Accept incoming connection
    ClientSocket := fpAccept(ServerSocket, nil, nil);
    if ClientSocket = -1 then
    begin
      // Ignore interrupted system call errors, log others
      if GetLastOSError <> ESysEINTR then
        WriteLn('Error accepting connection: ', GetLastOSError);
      Continue;
    end;

    WriteLn('Client connected.');
    // Handle client synchronously (simple echo)
    HandleClient(ClientSocket);
    WriteLn('Client disconnected.');
  end;

  // Close server socket (unreachable here)
  CloseSocket(ServerSocket);
end.
